Cookbook/Proposals/AdditionalRulesForDanger.md

```diff
@@ -0,0 +1,133 @@
+# Additional rules for Danger and SwiftLint
+
+* Author: João Pereira
+* Review Manager: TBD
+
+## Introduction
+
+The goal of this proposal is to introduce several new rules for Danger and/or SwiftLint. These will hopefully raise awareness for a number of problems we frequently see in our day-to-day work.
+
+## Motivation
+
+We have a large and particularly opinionated codebase (one could say "avant-gard" even) with a significant amount of deprecated code to address. New engineers face a somewhat steep learning curve before mastering our practices and the fact that we have at times two or three ways of addressing a particular problem (mostly due to legacy code) only brings more confusion to the mix.
+
+As such, at times it can ben confusing for newcomers to know what to do and what to avoid. Also, even for experienced engineers make mistakes so this proposal aims to lower the cognitive workload by making the rules more explicit and by improving automated issue detection. SwiftLint and Danger are two invaluable tools that we can and should continue to leverage in order to achieve these goals.  All of this will help prevent problems before these reach the `develop` branch. 
+
+The rules I am about to introduce are already known to us - we apply them on our PR reviews on a daily basis, after all - but, nevertheless, why rely on memory alone when we can automate them?
+
+## Proposed solution
+
+We will now go through the list of suggested rules and consider:
+    a) if they should be applied to our codebase (at all)
+    b) which tool (therefore, which approach) should we use to minimize further occurrences of the problem at hand
+
+### Custom ViewModels in the test suite
+
+Custom ViewModels should not be defined as part of our tests; we should rely on `StubViewModel`s (from `BabylonSnapshotTestUtilities`) instead (although, in theory, setting up custom ViewModels could be useful in particular circumstances).
```

Comment:
I don't have much context here. Why and what are these? In which circumstances can they be useful?

Reply:
We used to define stubs for view models in the snapshot tests as they are required to create the screens. Now we have a generic `StubViewModel` for that and we don't need custom implementations, but they are still being used somewhere. I'd prefer this to be solved with deprecation annotation though. We can for example write a variant of `verifyScreenForAllSizesAndVisualLanguages` that would only accept `StubViewModel` for `viewModel` parameter and deprecate the variant that is generic over type of ViewModel. @NSMyself  what do you think?

Reply:
@chitrakotwani apologies for not replying, I couldn't do it at the time and completely forgot about it. 

Basically, by using custom ViewModels instead of stubbing the ones we already have technically we aren't testing the actual ViewModels, but the ones we've just created for the purpose of this test. Depending on the actual implementation, that may not be the exact same thing.

Regarding the circumstances in which this approach could be useful, well:
a) we may need to test something that because of it's nature and/or complexity is not easily achievable and, as such, using a custom ViewModel could theoretically help (although that fact could indicate that the VM needs to be refactored, but that's another story).

b) a concrete example: when testing a Renderer and we need our VM to hold some value that we cannot define via the initializers and/or easily get the state machine to the point where we need it to be.

Reply:
@ilyapuchka i think that's a great idea. I need to revisit this proposal and update it because in some cases marking the method via a deprecation annotation is indeed the right approach imho. Will do so in the following days.